### PR TITLE
Workspace restart button/kira pilot

### PR DESCRIPTION
**Pull request by Kira-Pilot**
Trying to address #5800 and #6241
 
Need some help figuring out the best approach here. We've added a new transition, `restart` and attempted to amend `api.postWorkspaceBuilds` such that if a `restart` is requested, we insert two workspace builds: 1 for `stop` and 1 for `start`.

I'm worried Provisioner isn't set up to handle this scenario. @mafredri pointed out that if we create two jobs, we may have separate daemons execute each simultaneously.

### CODE REVIEW

File: coderd/workspacebuilds.go [hunk 1]
```diff
@@ -288,28 +288,10 @@ func (api *API) workspaceBuildByBuildNumber(rw http.ResponseWriter, r *http.Requ
...
-// @Accept json
-// @Produce json
-// @Tags Builds
-// @Param workspace path string true "Workspace ID" format(uuid)
-// @Param request body codersdk.CreateWorkspaceBuildRequest true "Create workspace build request"
-// @Success 200 {object} codersdk.WorkspaceBuild
-// @Router /workspaces/{workspace}/builds [post]
-// nolint:gocyclo
-func (api *API) postWorkspaceBuilds(rw http.ResponseWriter, r *http.Request) {
+func (api *API) postBuild(rw http.ResponseWriter, r *http.Request, createBuild codersdk.CreateWorkspaceBuildRequest) (codersdk.WorkspaceBuild, error) {
```
Comment (Kira-Pilot) [inline coderd/workspacebuilds.go:291]: In an effort not to duplicate logic from `postWorkspaceBuilds`, I've extracted the meat of that function and stuck it in this new helper, `postBuild`, which is then called from `postWorkspaceBuilds` and the new handler, `restartWorkspace`. I'm wondering if there's a better way to extract this logic. The way I'm handling errors (basically logging twice) seems stylistically incorrect.

File: coderd/workspacebuilds.go [hunk 2]
```diff
@@ -673,12 +660,84 @@ func (api *API) postWorkspaceBuilds(rw http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Internal error stopping workspace.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+
+	// create a build - start the workspace
+	build, err = api.postBuild(rw, r, codersdk.CreateWorkspaceBuildRequest{
```
Comment (Kira-Pilot) [inline coderd/workspacebuilds.go:698]: We seem to be starting the workspace before it has been fully stopped. How should I go about fixing this?

Comment (Kira-Pilot) [reply coderd/workspacebuilds.go:698]: Potentially related:  I notice in `cli/restart.go` we look for a [ProvisionerJob error](https://github.com/coder/coder/blob/main/cli/restart.go#L48); however, I'm unsure why we're doing that or how I should replicate that behavior inside this handler.

File: coderd/workspacebuilds.go [hunk 3]
```diff
@@ -673,12 +660,84 @@ func (api *API) postWorkspaceBuilds(rw http.ResponseWriter, r *http.Request) {
...
+	})
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Internal error starting workspace.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+
+	httpapi.Write(ctx, rw, http.StatusOK, build)
```
Comment (Kira-Pilot) [inline coderd/workspacebuilds.go:709]: Does it make sense to return the `start` build here? Is there a better status?

**Review by mafredri — COMMENTED** [review]
@Kira-Pilot I think your approach here can work, but ultimately I wonder if we should take a slightly different one.

My thought is that we would introduce a new transition state for workspaces called `restart` (https://github.com/coder/coder/blob/4dd5d794120cb60c3b7fd0f78c73ec1b7fda8727/codersdk/workspacebuilds.go#L14-L20).

The motivation behind a new state is that right now, the restart action is dependent on the Coder API and can be interrupted. Consider if we issue `restart`, and begin by stopping. Now while the workspace is stopping, the coder server is restarted (or updated). Now the workspace would remain in the stopped state.

The new restart state would allow for the state to be fully registered so that we can ensure it's completion.

Do you have any thoughts on this @kylecarbs?

**Review by mtojek — COMMENTED** [review]
Side note: the abort transaction error mentioned in the Github issue can be fixed by increasing the build number for the "start" job.

Regarding the overall design, the "restart" transition is purely virtual, and could do the trick if the provisionerserver doesn't acquire jobs that are conflicting. In this case, `stop` and `start` are executed at the same time:

Screenshot 2023-04-13 at 10 47 48

and the "stop" job fails:

Screenshot 2023-04-13 at 10 48 19

... or does nothing:

Screenshot 2023-04-13 at 10 50 49

Fun fact:

a similar inconsistency can be observed for the "start" job. It can just add a new resource:

Screenshot 2023-04-13 at 10 52 20

... or add the resource and destroy the old one (depending on the race with "stop" job):

Screenshot 2023-04-13 at 10 53 12

This inconsistency is kind of funny, because for some (all?) providers like Docker it's absolutely fine to just call `/workspacebuilds` with `transition=start`. I'm afraid that to fully solve the racing challenge, you need to implement mutual exclusiveness between provisioner jobs, so tweak the provisionerserver. I strongly recommend doing it in a separate PR. 

@mafredri Please doublecheck if this all makes sense 👍

**Comment by mafredri** [comment]
@mtojek great insights. I like your proposal of limiting the acquirement of conflicting jobs in the provisioner(s). One thing that worries me though is if `stop` fails, we should probably abort the `start` job? But thins brings me to another idea for potentially solving the conflict:

Provisioner job dependencies. I.e. a job could depend on another job. Meaning we wouldn't pick up a job if it's dependent on a job that hasn't completed. A toggle for requiring success or simply completion (to continue after fail) of the dependent job could also be added. (Motivation for toggle: Possibility to implement "force restart".)

**Comment by mtojek** [comment]
> Provisioner job dependencies. I.e. a job could depend on another job. Meaning we wouldn't pick up a job if it's dependent on a job that hasn't completed. A toggle for requiring success or simply completion (to continue after fail) of the dependent job could also be added. (Motivation for toggle: Possibility to implement "force restart".)

It is a valid point and good motivation for feature expansion. I'm a bit concerned about the provisioner job being left in the queue forever (if the dependent job failed), so maybe provisionerserver can pick it up and simply discard it.

**Comment by mafredri** [comment]
> It is a valid point and good motivation for feature expansion. I'm a bit concerned about the provisioner job being left in the queue forever (if the dependent job failed), so maybe provisionerserver can pick it up and simply discard it.

Yeah, that's about how I'd imagine it would behave as well. Or rather than discard, mark the dependent job as failed too.

**Comment by mafredri** [comment]
@Kira-Pilot sounds good, that is what the CLI does currently too so we should deal with it at some point but doesn't have to be now. 👍🏻

**Pull request closed without merging**